source-academy/js-slang
# Add Source 2 Typed Variant

Changes:
-
- Added support for `null` to type error checker.
- Added `Pair` and `List` types to type error checker.
- Added type checking for Source 2 library functions (e.g. `pair`, `list`, `head`, `tail`).
- Added tests for Source 2 Typed.

How to Test this PR:
-
1. Without frontend: Run Source 2 Typed in the js-slang REPL.
```
$ js-slang -c 2 -v typed # if js-slang is added to PATH
$ node dist/repl/repl.js -c 2 -v typed # if js-slang is not added to PATH
```
2. With frontend: [Link local js-slang to local frontend](https://github.com/source-academy/js-slang#using-your-js-slang-in-your-local-source-academy) and checkout local frontend to source-academy/frontend#2293.
3. Check the existing unit tests in `src/typeChecker/__tests__/source2Typed.test.ts`.
```
$ yarn test src/typeChecker/__tests__/source2Typed.test.ts
```
